src/librustc/hir/lowering.rs

```diff
@@ -68,15 +68,19 @@ pub struct LoweringContext<'a> {
     // the form of a DefIndex) so that if we create a new node which introduces
     // a definition, then we can properly create the def id.
     parent_def: Cell<Option<DefIndex>>,
-    resolver: Option<RefCell<&'a mut Resolver>>,
+    resolver: RefCell<&'a mut Resolver>,
 }
 
 pub trait Resolver {
+    // Resolve a global hir path generated by the lowerer when expanding `for`, `if let`, etc.
     fn resolve_generated_global_path(&mut self, path: &hir::Path, is_value: bool) -> Def;
 
-    fn def_map(&mut self) -> &mut DefMap;
+    // Record the resolution of a path or binding generated by the lowerer when expanding.
+    fn record_resolution(&mut self, id: NodeId, def: Def);
+
     // We must keep the set of definitions up to date as we add nodes that weren't in the AST.
-    fn definitions(&mut self) -> &mut Definitions;
+    // This should only return `None` during testing.
+    fn definitions(&mut self) -> Option<&mut Definitions>;
```

Comment:
I could further refactor this into `local_def_id` (which would return a dummy `DefId` during testing) and `crate_def_with_parent` (which would do nothing during testing).